---
title: try-catch を使用して例外を処理する方法 - C# プログラミング ガイド
description: try-catch ブロックを使用して例外を処理する方法について説明します。 コード例を参照し、使用可能なその他のリソースを確認します。
ms.date: 12/09/2020
helpviewer_keywords:
- exception handling [C#], try/catch blocks
- exceptions [C#], try/catch blocks
- try/catch blocks [C#]
ms.assetid: ca8e3773-980e-4767-8633-7408540e9818
ms.openlocfilehash: b6368660dbe037123f5bb6ce52502d4a94fcfc3a
ms.sourcegitcommit: 9b877e160c326577e8aa5ead22a937110d80fa44
ms.translationtype: HT
ms.contentlocale: ja-JP
ms.lasthandoff: 12/11/2020
ms.locfileid: "97110521"
---
# <a name="how-to-handle-an-exception-using-trycatch-c-programming-guide"></a><span data-ttu-id="411d6-104">try/catch を使用して例外を処理する方法 (C# プログラミング ガイド)</span><span class="sxs-lookup"><span data-stu-id="411d6-104">How to handle an exception using try/catch (C# Programming Guide)</span></span>

<span data-ttu-id="411d6-105">[try-catch](../../language-reference/keywords/try-catch.md) ブロックの目的は、作業コードによって生成された例外をキャッチし、処理することです。</span><span class="sxs-lookup"><span data-stu-id="411d6-105">The purpose of a [try-catch](../../language-reference/keywords/try-catch.md) block is to catch and handle an exception generated by working code.</span></span> <span data-ttu-id="411d6-106">例外によっては、`catch` ブロックで処理され、例外を再スローせずに問題を解決できるものもありますが、多くの場合、適切な例外が確実にスローされるようにする必要があります。</span><span class="sxs-lookup"><span data-stu-id="411d6-106">Some exceptions can be handled in a `catch` block and the problem solved without the exception being rethrown; however, more often the only thing that you can do is make sure that the appropriate exception is thrown.</span></span>

## <a name="example"></a><span data-ttu-id="411d6-107">例</span><span class="sxs-lookup"><span data-stu-id="411d6-107">Example</span></span>

<span data-ttu-id="411d6-108">この例の <xref:System.IndexOutOfRangeException> は最も適切な例外というわけではありません。呼び出し元から `index` 引数が渡されてエラーが発生しているため、より適切なメソッドは <xref:System.ArgumentOutOfRangeException> になります。</span><span class="sxs-lookup"><span data-stu-id="411d6-108">In this example, <xref:System.IndexOutOfRangeException> isn't the most appropriate exception: <xref:System.ArgumentOutOfRangeException> makes more sense for the method because the error is caused by the `index` argument passed in by the caller.</span></span>

:::code language="csharp" source="snippets/exceptions/ExampleTryCatch.cs" id="ExampleTryCatch":::

## <a name="comments"></a><span data-ttu-id="411d6-109">コメント</span><span class="sxs-lookup"><span data-stu-id="411d6-109">Comments</span></span>

<span data-ttu-id="411d6-110">例外を発生させるコードは `try` ブロックに囲まれています。</span><span class="sxs-lookup"><span data-stu-id="411d6-110">The code that causes an exception is enclosed in the `try` block.</span></span> <span data-ttu-id="411d6-111">`IndexOutOfRangeException` が発生した場合にこれを処理するための `catch` ステートメントが、すぐ後に追加されています。</span><span class="sxs-lookup"><span data-stu-id="411d6-111">A `catch` statement is added immediately after to handle `IndexOutOfRangeException`, if it occurs.</span></span> <span data-ttu-id="411d6-112">`catch` ブロックは `IndexOutOfRangeException` を処理し、代わりにより適切な `ArgumentOutOfRangeException` 例外をスローします。</span><span class="sxs-lookup"><span data-stu-id="411d6-112">The `catch` block handles the `IndexOutOfRangeException` and throws the more appropriate `ArgumentOutOfRangeException` exception instead.</span></span> <span data-ttu-id="411d6-113">呼び出し元にできるだけ多くの情報を提供するため、元の例外を新しい例外の <xref:System.Exception.InnerException%2A> として指定することを検討してください。</span><span class="sxs-lookup"><span data-stu-id="411d6-113">In order to provide the caller with as much information as possible, consider specifying the original exception as the <xref:System.Exception.InnerException%2A> of the new exception.</span></span> <span data-ttu-id="411d6-114"><xref:System.Exception.InnerException%2A> プロパティは[読み取り専用](../../properties.md#read-only)であるため、新しい例外のコンストラクターで割り当てる必要があります。</span><span class="sxs-lookup"><span data-stu-id="411d6-114">Because the <xref:System.Exception.InnerException%2A> property is [read-only](../../properties.md#read-only), you must assign it in the constructor of the new exception.</span></span>
